### PR TITLE
[tune] HyperOpt Support (v2)

### PR DESCRIPTION
**Pull request by richardliaw**
- [x] Tests:
   - [x] Fix jenkins by creating local docker image of jenkins env
 - [x] Clarify comment for verbose
 - [x] Docs
  - [x] Document necessity for this `pip install --upgrade git+git://github.com/hyperopt/hyperopt.git`

### CODE REVIEW

**Comment by richardliaw** [comment]
screenshot 2018-03-30 22 21 36

File: python/ray/tune/experiment.py [hunk 1]
```diff
@@ -49,8 +48,13 @@ def __init__(self, name, run, stop=None, config=None,
...
-        self._trials = generate_trials(spec, name)
 
-    def trials(self):
-        for trial in self._trials:
-            yield trial
+        self.name = name
+        self.spec = spec
+
+
+class JSONExperiment(Experiment):
```
Comment (ericl) [inline python/ray/tune/experiment.py]: Can this be a function that returns an Experiment object instead, e.g.,

Experiment.from_json(name, spec)

File: python/ray/tune/hpo_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+from __future__ import division
+from __future__ import print_function
+
+import os
+import copy
+import numpy as np
+try:
+    import hyperopt as hpo
+except Exception as e:
+    print("HyperOpt not found; perhaps not installed?")
```
Comment (ericl) [inline python/ray/tune/hpo_scheduler.py]: If you want to make this a soft dep, set hpo = None, and check this in the HyperOptScheduler constructor

File: python/ray/tune/hpo_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+from ray.tune.trial import Trial
+from ray.tune.error import TuneError
+from ray.tune.registry import register_trainable
+from ray.tune.trial_scheduler import TrialScheduler, FIFOScheduler
+from ray.tune.config_parser import make_parser
+from ray.tune.variant_generator import to_argv
+
+
+class HyperOptScheduler(FIFOScheduler):
```
Comment (ericl) [inline python/ray/tune/hpo_scheduler.py]: class doc

File: python/ray/tune/hpo_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+    def __init__(self, max_concurrent=10, loss_attr="mean_loss"):
+        self._max_concurrent = max_concurrent  # NOTE: this is modified later
```
Comment (ericl) [inline python/ray/tune/hpo_scheduler.py]: Shouldn't we always launch as many trials at once as resources permits?

File: python/ray/tune/hpo_scheduler.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        self._loss_attr = loss_attr
+        self._experiment = None
+
+    def track_experiment(self, experiment, trial_runner):
```
Comment (ericl) [inline python/ray/tune/hpo_scheduler.py]: add_experiment()

File: python/ray/tune/hpo_scheduler.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        return self._num_trials_left > 0
+
+    def get_hyperopt_trials(self):
+        return self._hpopt_trials
+
+    def choose_trial_to_run(self, trial_runner):
+        self.refresh_queue(trial_runner)
+        return FIFOScheduler.choose_trial_to_run(self, trial_runner)
+
+    def refresh_queue(self, trial_runner):
```
Comment (ericl) [inline python/ray/tune/hpo_scheduler.py]: _add_new_trials_if_needed()

File: python/ray/tune/hpo_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        return {"loss": getattr(result, self._loss_attr),
+                "status": "ok"}
+
+    def _get_dynamic_trial(self, tid):
+        return [t for t in self._hpopt_trials.trials if t["tid"] == tid][0]
+
+    def _continue(self):
+        return self._num_trials_left > 0
+
+    def get_hyperopt_trials(self):
```
Comment (ericl) [inline python/ray/tune/hpo_scheduler.py]: unused method?

File: python/ray/tune/hpo_scheduler.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+        self.refresh_queue(trial_runner)
+        return FIFOScheduler.choose_trial_to_run(self, trial_runner)
+
+    def refresh_queue(self, trial_runner):
+        """Checks if there is a next trial ready to be queued.
+
+        This is determined by tracking the number of concurrent
+        experiments and trials left to run."""
+        while (self._num_trials_left > 0 and
+               self._num_live_trials() < self._max_concurrent):
```
Comment (ericl) [inline python/ray/tune/hpo_scheduler.py]: You can also check if there does not exist any PENDING trials. That means we can add another one.

File: python/ray/tune/hpo_scheduler.py [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+        hp_result = self._convert_result(result)
+        ho_trial['result'] = hp_result
+        self._hpopt_trials.refresh()
+        del self._tune_to_hp[trial]
+
+    def _convert_result(self, result):
+        return {"loss": getattr(result, self._loss_attr),
+                "status": "ok"}
+
+    def _get_dynamic_trial(self, tid):
```
Comment (ericl) [inline python/ray/tune/hpo_scheduler.py]: _get_hyperopt_trial

Also can't we do a dict lookup here?

File: python/ray/tune/hpo_scheduler.py [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+    def on_trial_complete(self, trial_runner, trial, result):
+        ho_trial = self._get_dynamic_trial(self._tune_to_hp[trial])
+        ho_trial['refresh_time'] = hpo.utils.coarse_utcnow()
+        ho_trial['state'] = hpo.base.JOB_STATE_DONE
+        hp_result = self._convert_result(result)
+        ho_trial['result'] = hp_result
+        self._hpopt_trials.refresh()
+        del self._tune_to_hp[trial]
+
+    def _convert_result(self, result):
```
Comment (ericl) [inline python/ray/tune/hpo_scheduler.py]: _to_hyperopt_result

File: python/ray/tune/hpo_scheduler.py [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+        return {"loss": getattr(result, self._loss_attr),
+                "status": "ok"}
+
+    def _get_dynamic_trial(self, tid):
+        return [t for t in self._hpopt_trials.trials if t["tid"] == tid][0]
+
+    def _continue(self):
```
Comment (ericl) [inline python/ray/tune/hpo_scheduler.py]: unused?

File: python/ray/tune/hpo_scheduler.py [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+        experiments and trials left to run."""
+        while (self._num_trials_left > 0 and
+               self._num_live_trials() < self._max_concurrent):
+            try:
+                trial_runner.add_trial(next(self.trial_generator))
+            except StopIteration:
+                break
+
+    def _num_live_trials(self):
+        return len(self._tune_to_hp)
```
Comment (ericl) [inline python/ray/tune/hpo_scheduler.py]: can inline this

File: python/ray/tune/trial.py [hunk 13]
```diff
@@ -333,8 +333,8 @@ def restore_from_obj(self, obj):
     def update_last_result(self, result, terminate=False):
         if terminate:
             result = result._replace(done=True)
-        if terminate or (
-                self.verbose and
+        if self.verbose and (
```
Comment (ericl) [inline python/ray/tune/trial.py]: shouldn't we always print the result if it terminated?

File: python/ray/tune/trial_runner.py [hunk 14]
```diff
@@ -40,13 +40,15 @@ class TrialRunner(object):
...
-                 server_port=TuneServer.DEFAULT_PORT):
+                 server_port=TuneServer.DEFAULT_PORT, verbose=True):
         """Initializes a new TrialRunner.
 
         Args:
             scheduler (TrialScheduler): Defaults to FIFOScheduler.
             launch_web_server (bool): Flag for starting TuneServer
-            server_port (int): Port number for launching TuneServer"""
+            server_port (int): Port number for launching TuneServer
+            verbose (bool): How much output should be printed for each trial.
```
Comment (ericl) [inline python/ray/tune/trial_runner.py]: This comment should be clarified

File: python/ray/tune/trial_scheduler.py [hunk 15]
```diff
@@ -64,6 +72,14 @@ def debug_string(self):
 class FIFOScheduler(TrialScheduler):
     """Simple scheduler that just runs trials in submission order."""
 
+    def track_experiment(self, experiment, trial_runner):
+        generator = generate_trials(experiment.spec, experiment.name)
+        while True:
+            try:
+                trial_runner.add_trial(next(generator))
+            except StopIteration:
+                break
```
Comment (ericl) [inline python/ray/tune/trial_scheduler.py]: Move this into TrialScheduler?

File: python/ray/tune/trial_scheduler.py [hunk 16]
```diff
@@ -47,11 +48,18 @@ def on_trial_remove(self, trial_runner, trial):
 
         raise NotImplementedError
 
-    def choose_trial_to_run(self, trial_runner, trials):
+    def track_experiment(self, experiment, trial_runner):
```
Comment (ericl) [inline python/ray/tune/trial_scheduler.py]: `add_experiment`

File: python/ray/tune/trial_scheduler.py [hunk 17]
```diff
@@ -47,11 +48,18 @@ def on_trial_remove(self, trial_runner, trial):
...
+        """Tracks experiment from which trials and generated and queued."""
```
Comment (ericl) [inline python/ray/tune/trial_scheduler.py]: """Adds an experiment to the scheduler. The scheduler is responsible for adding the trials of the experiment to the runner, which can be done immediately (if there are a finite set of trials), or over time (if there is an infinite stream of trials)."""

File: python/ray/tune/trial_scheduler.py [hunk 18]
```diff
@@ -47,11 +48,18 @@ def on_trial_remove(self, trial_runner, trial):
...
+
+        raise NotImplementedError
+
+    def choose_trial_to_run(self, trial_runner):
         """Called to choose a new trial to run.
 
         This should return one of the trials in trial_runner that is in
-        the PENDING or PAUSED state."""
+        the PENDING or PAUSED state. This is assumed to be idempotent.
```
Comment (ericl) [inline python/ray/tune/trial_scheduler.py]: This function must be idempotent.

File: python/ray/tune/tune.py [hunk 19]
```diff
@@ -4,28 +4,49 @@
...
-    "HyperBand": HyperBandScheduler,
-    "AsyncHyperBand": AsyncHyperBandScheduler,
-}
+def import_schedulers():
+    for key in ["FIFO", "MedianStopping", "HyperBand",
+                "AsyncHyperBand", "HyperOpt"]:
+        try:
+            _SCHEDULERS[key] = get_scheduler(key)
+        except ImportError as e:
+            print("Could not import scheduler: {}".format(key))
```
Comment (ericl) [inline python/ray/tune/tune.py]: If this is just for HyperOpt, let's just have that fail in the constructor instead of on import.

**Review by ericl — CHANGES_REQUESTED** [review]
Looking good, a bunch of clarification comments.

File: python/ray/tune/hpo_scheduler.py (hunk 4 above)
Comment (richardliaw) [reply python/ray/tune/hpo_scheduler.py]: Not necessarily; the more concurrent trials the less effective your search. For example, SigOpt only allows for 10 concurrent trials.

Now, for this particular case it could make sense to default to as many as resource permits.

File: python/ray/tune/hpo_scheduler.py (hunk 9 above)
Comment (richardliaw) [reply python/ray/tune/hpo_scheduler.py]: I would prefer not to since hyperopt is pretty fragile (this exact trial dictionary that we get from this method is not a copy but rather the actual object used in trial generation)

It wouldn't be horrible + we can write a test but one could imagine it as a source for error

File: python/ray/tune/hpo_scheduler.py (hunk 12 above)
Comment (richardliaw) [reply python/ray/tune/hpo_scheduler.py]: just tried it out and it's a lot less readable

File: python/ray/tune/trial.py (hunk 13 above)
Comment (richardliaw) [reply python/ray/tune/trial.py]: in testing out HyperOpt, I wanted a way to only get the debugstr (that listed all the trials at once) as it's clean and it's the only thing I needed. 

Even when printing out only results of terminated trials, stdout would be flooded.

File: python/ray/tune/trial_scheduler.py (hunk 15 above)
Comment (richardliaw) [reply python/ray/tune/trial_scheduler.py]: no strong opinion about this; is there a benefit to keeping TrialScheduler as an interface?

File: python/ray/tune/hpo_scheduler.py (hunk 3 above)
Comment (richardliaw) [reply python/ray/tune/hpo_scheduler.py]: done

**Comment by richardliaw** [comment]
OK for testing, HyperOpt only works if built from source. Do we want to do this every build? Alternatively we could host a wheel somewhere, or treat it like PyTorch where we don't test this codepath.

File: python/ray/tune/hpo_scheduler.py [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+from ray.tune.config_parser import make_parser
+from ray.tune.variant_generator import to_argv
+
+
+class HyperOptScheduler(FIFOScheduler):
+    """FIFOScheduler that uses HyperOpt to provide trial suggestions.
+
+    Requires HyperOpt to be installed. Uses the Tree of Parzen Estimators
+    algorithm. Note that this class takes in a loss attribute
+    rather than a reward.
```
Comment (ericl) [inline python/ray/tune/hpo_scheduler.py]: Should we change all of them to loss attribute?

File: python/ray/tune/hpo_scheduler.py [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+
+    Parameters:
+        max_concurrent (int | None): Number of maximum concurrent trials.
+            If None, then trials will be queued only if resources
+            are available.
+        loss_attr (str): The TrainingResult objective value attribute.
```
Comment (ericl) [inline python/ray/tune/hpo_scheduler.py]: If you use neg_mean_loss then it can be any increasing attribute.

File: python/ray/tune/hpo_scheduler.py [hunk 22]
```diff
@@ -0,0 +1,189 @@
...
+                new_ids, self.domain, self._hpopt_trials,
+                self.rstate.randint(2 ** 31 - 1))
+            self._hpopt_trials.insert_trial_docs(new_trials)
+            self._hpopt_trials.refresh()
+            new_trial = new_trials[0]
+            new_trial_id = new_trial["tid"]
+            suggested_config = hpo.base.spec_from_misc(new_trial["misc"])
+            new_cfg.update(suggested_config)
+
+            kv_str = "_".join(["{}={}".format(k, str(v)[:5])
```
Comment (ericl) [inline python/ray/tune/hpo_scheduler.py]: Consider sorted() to produce a deterministic order

File: python/ray/tune/hpo_scheduler.py [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+                               for k, v in suggested_config.items()])
+            experiment_tag = "hyperopt_{}_{}".format(new_trial_id, kv_str)
```
Comment (ericl) [inline python/ray/tune/hpo_scheduler.py]: Drop the hyperopt prefix?

File: python/ray/tune/trial.py (hunk 13 above)
Comment (ericl) [reply python/ray/tune/trial.py]: Ic, let's try it then.

File: python/ray/tune/trial_scheduler.py (hunk 15 above)
Comment (ericl) [reply python/ray/tune/trial_scheduler.py]: It doesn't seem particularly specific to fifo scheduler. I imagine most schedulers will want to extend this instead of fifoscheduler.

Speaking of which should HyperOpt extend the scheduler interface instead of fifoscheduler?

**Comment by ericl** [comment]
Hm why does it have to be built from source? If it's fast then it seems ok.

**Comment by richardliaw** [comment]
Due to incompatibilities with networkx, see https://github.com/hyperopt/hyperopt/issues/325

File: python/ray/tune/hpo_scheduler.py (hunk 20 above)
Comment (richardliaw) [reply python/ray/tune/hpo_scheduler.py]: Hm I guess we could, but in separate PR

File: python/ray/tune/trial_scheduler.py (hunk 15 above)
Comment (richardliaw) [reply python/ray/tune/trial_scheduler.py]: I think wrt extending HyperOpt it's fine to extend fifo, as it does call `FIFOScheduler` methods for `on_trial_add` and also for `choose_trial_to_run`. 

That being said, the `choose_trial_to_run` invocation is done as a weird class method invocation rather than a super invocation and could work either way, so I'm not married to anything...

File: python/ray/tune/hpo_scheduler.py (hunk 20 above)
Comment (ericl) [reply python/ray/tune/hpo_scheduler.py]: I think we should still to `reward_attr` for consistency then (basically negate it in this pr). We can fix up / have a more easy to understand terminology in the followup.

**Comment by ericl** [comment]
Is it possible to pip install the github URL in the tests?

**Comment by richardliaw** [comment]
OK fixed it so that a CI example test runs; how strongly do you feel about unit testing this? Perhaps we can just mock the hyperopt for Travis instead of installing from source for all tests

**Comment by richardliaw** [comment]
retest this please

**Comment by richardliaw** [comment]
retest this please

**Comment by richardliaw** [comment]
OK interestingly enough importing hyperopt before Ray tune causes
tensorflow to fail to import ...

**Review by ericl — APPROVED** [review]
LGTM

**Comment by richardliaw** [comment]
retest this please

**Comment by richardliaw** [comment]
retest this please

**Comment by richardliaw** [comment]
retest this please

**Comment by richardliaw** [comment]
retest this please

**Comment by richardliaw** [comment]
retest this please

**Comment by richardliaw** [comment]
retest this please

**Comment by richardliaw** [comment]
travis completely (!) passes on private branch, merging!

This also subsumes #1823.

**Comment by robertnishihara** [comment]
What! This is unheard of :)
